File: src/main/java/cspfj/generator/constraint/GccGenerator.java
package cspfj.generator.constraint;

import java.util.List;

import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import cspfj.constraint.semantic.Gcc;
import cspfj.constraint.semantic.Gcc.Bounds;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.GeneralConstraint;

public class GccGenerator extends AbstractGenerator {

    public GccGenerator(Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        if (!(constraint instanceof GeneralConstraint)) {
            throw new FailedGenerationException(constraint
                    + " is not supported");
        }

        final List<Variable> scope = Lists.transform(constraint.getScope(),
                CSPOM_TO_CSP4J);
        if (Iterables.any(scope, NULL_DOMAIN)) {
            return false;
        }

        final String[] params = constraint.getParameters().split(", +");

        if (params.length % 3 != 0) {
            throw new FailedGenerationException("3 parameters per value");
        }

        final Bounds[] bounds = new Bounds[params.length / 3];
        for (int i = 0; i < params.length / 3; i++) {
            bounds[i] = new Bounds(Integer.valueOf(params[3 * i]),
                    Integer.valueOf(params[3 * i + 1]),
                    Integer.valueOf(params[3 * i + 2]));
        }

        addConstraint(new Gcc(scope.toArray(new Variable[scope.size()]), bounds));
        return true;
    }

}

File: src/main/scala/cspfj/generator/constraint/GccGenerator.scala
package cspfj.generator.constraint;

import cspfj.constraint.semantic.Bounds
import cspfj.constraint.semantic.Gcc
import cspfj.problem.{ Variable, Problem }
import cspom.constraint.{ GeneralConstraint, CSPOMConstraint }

final class GccGenerator(problem: Problem) extends AbstractGenerator(problem) {
  def generate(constraint: CSPOMConstraint) = {
    require(constraint.isInstanceOf[GeneralConstraint])

    val scope = constraint.scope map cspom2cspfj

    if (scope exists (_.dom == null)) {
      false
    } else {
      val params = constraint.asInstanceOf[GeneralConstraint].predicate.parameters.get.split(", +");

      require(params.size % 3 == 0, "3 parameters per value");

      val bounds = params.grouped(3) map (p => new Bounds(p(0).toInt, p(1).toInt, p(2).toInt))

      addConstraint(new Gcc(scope.toArray, bounds.toArray));
      true;
    }

  }
}